spark/src/main/scala/geotrellis/spark/op/focal/FocalOperation.scala

```diff
@@ -10,160 +12,117 @@ import org.apache.spark.SparkContext._
 import spire.syntax.cfor._
 
 import annotation.tailrec
+import scala.reflect.ClassTag
 
 object FocalOperation {
 
-  private def getNeighborCoordinates(gridBounds: GridBounds, col: Int, row: Int) = {
-
-    val index = (row % 3) * 3 + (col % 3)
-
-    val neighborCoordinates = Array.ofDim[Option[(Int, Int)]](9)
-    cfor(0)(_ < 9, _ + 1) { i =>
-      val (dy, dx) = FocalOperation.indicesDifferenceToCoords(index, i)
-      val (r, c) = (row + dy, col + dx)
-
-      neighborCoordinates(i) =
-        if (c >= gridBounds.width || r >= gridBounds.height || c < 0 || r < 0) None
-        else Some((r, c))
-    }
-
-    neighborCoordinates.toSeq
-  }
-
-  private def indicesDifferenceToCoords(from: Int, to: Int) = {
-    val xStart = if (from % 3 == 0) 0 else if (from % 3 == 1) -1 else 1
-    val yStart = if (from / 3 == 0) 0 else if (from / 3 == 1) -1 else 1
-
-    val y = (yStart + to / 3) % 3 match {
-      case 2 => -1
-      case y => y
-    }
-
-    val x = (xStart + to % 3) % 3 match {
-      case 2 => -1
-      case x => x
+  object Direction extends Enumeration {
+    type Direction = Value
+    val Center, Top, TopRight, Right, BottomRight, Bottom, BottomLeft, Left, TopLeft = Value
+
+    implicit class withMethods(direction: Value) {
```

Comment:
Really good question. I've read that enums will be serialized better and took it on faith. Your question prompted me to test it and I could not get Kryo to serialize the num sensibly. It would always produces something like: `??scala.Enumeration$Va￬???geotrellis.spark.op.focal.FocalOperation$Directionﾤ?????`

Perhaps the statement applies to java enums?

I switched to sealed trait and was able to register them and get it to serialize into 2 byte id through kryo. It didn't make anything faster locally but I bet it would have some non-zero impact on network traffic